### PR TITLE
VACMS-9737: Implements Q&A reusability for Resource & Support

### LINKED ISSUE #9737: Drupal: Implement reusable Q&A Group in R&S Detail content type, with text-filter capability
## Description

As an Editor, I want to easily add reusable Q&As to a Q&A Group on Resources & Support Detail pages, so that Veterans get better content with less effort.

## Tasks
Remaining tasks, a draft PR exists, so work to get it over the line: 

- [x] Christia brain refresh - .5hr
- [x] Update [Airtable] - (https://airtable.com/appSg80z9nVWoesxn/tbl5o1AxKu7ufswjv/viwWTQYPMWSOV91rJ?blocks=bipz2F2iCJ6xlOlNM) for source of truth CMS fields. If no permissions, talk to Lynn - 1hr
    - [x] Behat tests for any new fields - ~2hrs~  These are generated by the spectool that was just updated ^^
- [x] GraphQL query in content-build repo, to add Description field - 2hrs
- [x] Code review / shepherding - 1-2hrs

6.5 hrs - 7.5hrs

## Acceptance Criteria
- [x] Add description ~paragraph~ field to Q&A Group
   - [ ] Help text provided/approved by CMS sitewide.
- [x] Make Q&A Group available on R&S Detail content type
- [x] Add text filtering capability to the Add Question modal so Editors can more easily find the reusable Q&A they want
- [x] Show both draft and published Q&As in Add Question modal, but hide archived ones
- [x] Provide GraphQL query / collaborate with FE eng 
- [x] Review permissions to ensure the new fields / QA group configs are editable by the same permission levels / Section editors as Q&A group on FAQ
- [ ] Test/QA
- [x] Requires design review
- [ ] Updates to related knowledge base article(s)  (publishing of the draft can be done in hook_update_n [with this function](https://github.com/department-of-veterans-affairs/va.gov-cms/blob/420d259ba0c8b4e0b52374b29628ae20a30f99e5/docroot/modules/custom/va_gov_db/va_gov_db.install#L239)  it goes out with this merge. 

### CMS Team
Please check the team(s) that will do this work.

- [ ] `Program`
- [ ] `Platform CMS Team`
- [ ] `Sitewide Crew`
- [ ] `⭐️ Sitewide CMS`
- [x] `⭐️ Public Websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

### PR DESCRIPTION
**Pull request by chri5tia**
Including added features.

* Added a limited rich text description field to Q&A Groups paragraph type

* Added Q&A Groups paragraph type to Resource and Support content type

* Edited Q&A Entity Browser view to include title/keyword filter,
  remove archived Q&A from populating and show answer preview

## Description
closes #9737
Relates to #9579, #8630

## Testing done
Hand-tested backend content add form.

## Screenshots

Screen Shot 2022-07-07 at 3 01 23 AM

Screen Shot 2022-07-07 at 6 07 59 PM

Screen Shot 2022-07-07 at 5 41 30 PM

Screen Shot 2022-07-07 at 1 04 49 AM

## QA steps

What needs to be checked to prove this works?  

- Editor can add or edit Resources and Support content, add main content block and choose Q&A Group for a modal to pop up (see above screenshots). Please edit node 47411 so you can run the GraphQL query I provided in t[his comment](https://github.com/department-of-veterans-affairs/va.gov-cms/pull/9746#issuecomment-1196292977) below.
   - go to https://pr9746-zqis1amj7i3jmmqvydvdxfmptbsn7zzf.ci.cms.va.gov/node/47411/edit 
   - add a QA group to main content
   - [x] validate that it appears and can be added. 
- [x] Content shown in modal should not include archived
- [x] Modal should have keyword/title filter option
- [x]  description field is not visible on node edit form
- [ ] ~Pattern change and new fields should be able to be pulled into frontend~ Covered by FE ticket https://github.com/department-of-veterans-affairs/content-build/pull/1253
- [ ] Accordion display works when chosen (not sure how to see this in action, only showed the questions before and after I made changes)

What needs to be checked to prove it didn't break any related things?  

- [x] Edit existing Resource and Support to include new Q&A Group content block
- [x] Edit or create FAQ content type and check that new Description field doesn't cause issues if it's not populated (it is not required)
- [x] Access for non-editors and editors should not have changed and should still be appropriate

What variations of circumstances (users, actions, values) need to be checked?  

- [ ] Non-editors can't edit
- [ ] Accordion type vs. non-accordion type should work the same way as it did before
- [ ] Q&A Group can be moved/rearranged on the page with respect to other main content blocks; old way of putting in Q&A content on a particular node can be edited/removed

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.
- [ ] Design approval
- [ ] FE isn't blocked by an issue/needed change

### Select Team for PR review

- [ ] `Platform CMS Team`
- [ ] `Sitewide program`
- [ ] `⭐️ Sitewide CMS`
- [x] `⭐️ Public websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`
- [ ]  Wait for FE related ticket to be good to go

### Does this PR need review from a Product Owner

- [ ] `Needs PO review` TBD

### CMS user-facing announcement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [x] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

File: config/sync/geocoder.geocoder_provider.mapbox.yml [hunk 1]
```diff
@@ -11,3 +11,5 @@ configuration:
     limit: null
   accessToken: 'set with environment variable'
   country: ''
+  location_type: address
+  fuzzy_match: false
```
Comment (swirtSJW) [inline config/sync/geocoder.geocoder_provider.mapbox.yml:15]: These changes seem unrelated.  My guess is they came from either an out of date db or code.  They should not be included in this pr.

**Review by swirtSJW — COMMENTED** [review]
This is looking like a great start.  
It'll need spectool tests updates to get the behat tests passing.
We also need to get some help text for the "Description" field written and approved by the sitewide CMS team.  I added both of these to the AC's on the ticket.

**Comment by chri5tia** [comment]
Edit node 47411 as follows:
- add content block > Q&A Group
- select 3 or so items and save
- Run this query in the cms tugboat instance:

```
query GetNodeSupportResourcesDetailPage($onlyPublishedContent: Boolean!) {
    nodeQuery(limit: 100, filter: {
      conditions: [
        { field: "status", value: ["1"], enabled: $onlyPublishedContent },
        { field: "type", value: ["support_resources_detail_page"] }
        { field: "nid", value: ["47411"]}
      ]
    }) {
      entities {
				... on NodeSupportResourcesDetailPage {
					fieldContentBlock {
					  targetId
					  targetRevisionId
            entity {
              entityId
              ... on ParagraphQAGroup {
                fieldSectionHeader
                fieldRichWysiwyg {
                  value
                  format
                  processed
                }
                queryFieldQAs {
                	entities {
                    entityId
                    entityLabel
                    moderationState
                    ... on NodeQA {
                      fieldAnswer {
                        targetId
                        targetRevisionId
                        ... on FieldNodeQAFieldAnswer {
                          entity {
                            entityId
                            entityLabel
                            
                          }
                        }
                      }
                    }
                  } 
                }
              }
            }
					}
        }
      }
    }
  }
```

File: config/sync/geocoder.geocoder_provider.mapbox.yml (hunk 1 above)
Comment (chri5tia) [reply config/sync/geocoder.geocoder_provider.mapbox.yml]: Removed, good catch!

**Comment by chri5tia** [comment]
-[ ] Update [Airtable] - (https://airtable.com/appSg80z9nVWoesxn/tbl5o1AxKu7ufswjv/viwWTQYPMWSOV91rJ?blocks=bipz2F2iCJ6xlOlNM) for source of truth CMS fields. If no permissions, talk to Lynn - 1hr

**Comment by chri5tia** [comment]
Check with UX and docs regarding the description field and other changes here.

**Comment by swirtSJW** [comment]
I converted this to draft so that it does not accidentally get merged before UX and FE solutions are in place.  We don't want a field going out and confusing editors because it has no visibility on the FE.
The other option is to go ahead with this but hide the field in the CMS until the FE is ready to use it.

**Comment by jilladams** [comment]
@swirtSJW we intend to merge the backend, frontend, and graphQL work in this sprint. In this case, the editors are Sitewide Content (rather than VAMC editors) and are ready to input content as soon as fields are available, so not sure the restriction to add logic to hide in CMS should apply in this case. Thoughts?

**Comment by swirtSJW** [comment]
> @swirtSJW we intend to merge the backend, frontend, and graphQL work in this sprint. In this case, the editors are Sitewide Content (rather than VAMC editors) and are ready to input content as soon as fields are available, so not sure the restriction to add logic to hide in CMS should apply in this case. (The front end PRs are just waiting to merge til the content model is merged for the sake of graphQL, for the sake of automated tests.) Thoughts?

It is not a hard and fast rule.  In general, we try not to fields that don't do anything exposed to editors as that creates a confusing user experience.   You could make the case that since there is a smaller number of editors in sitewide that it would be ok as you could communicate the change and the lag to them.  

Doing it as you describe would lag a full day.   CMS release at 3:30 the day before then FE templates going with the next day's build.

Assuming all this FE and CMS have been tested in a tugboat env then it could go like that.

The other option is releasing the CMS as hidden and unhiding it after the FE goes out (that is a really common pattern we use).

**Comment by jilladams** [comment]
Noted. In this case, I don't think we have any concerns about the lag, and Danielle / her team are looped in on these developments. But: for future, are there any docs around "releasing the CMS as hidden and unhiding it"? I've heard of that sort of thing via feature flags in Vets API but not in the context of the CMS itself, unless that's what you mean here, too.

**Comment by swirtSJW** [comment]
> Noted. In this case, I don't think we have any concerns about the lag, and Danielle / her team are looped in on these developments. But: for future, are there any docs around "releasing the CMS as hidden and unhiding it"? I've heard of that sort of thing via feature flags/flipper in Vets API but not in the context of the CMS itself, unless that's what you mean here, too.

There are [feature flags in the CMS](https://github.com/department-of-veterans-affairs/va.gov-cms/blob/main/READMES/interfaces.md#featureflags) that sync to the FE that could be used but would not work in this case.  When the content-build repos was split out from the vets-website build the CMS feature flags connection were broken  where the feature flags are no longer available during the query process. It used to be that we could say "If this feature flag is on, use query A, if it is off, use query B."   Currently the CMS feature flags are not available query time, they only become available after the queries.  For similar reasons, flipper flags are not available at the query level either.

Even if they were  available, they would be overkill in this case.  This is a simple matter of field visibility, which is something that is basic to Drupal and the flow kind of looks like this in terms of merge order:
1. Day 1: CMS PR creates the field as hidden.
2. Day 2: FE PR adds the query and template changes (templates are defensive in that they don't output the new content if it is not present)
3. Day 3: CMS PR 
   - makes the field visible
   - auto publishes latest draft of Knowledgebase Article(s
4. Notify editors after CMS build completes.

With that model, there is no editor confusion window.

Otherwise we have the compressed version:
1. Day 1: CMS PR creates the field visible, and publishes draft of Knowledgebase article(s)
2. . Notify editors after CMS build completes.
3. Day 2: FE PR adds the query and template changes (templates are defensive in that they don't output the new content if it is not present)

This model has a editor confusion window of at least 23 hours if all releases go as scheduled.

**Comment by swirtSJW** [comment]
Actually, I was wrong when I said this documentation doesn't exist.   It does.  https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/cms/product-team-support/content-model-changes.md

**Comment by swirtSJW** [comment]
I just added the abbreviated versions to that doc  https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/cms/product-team-support/content-model-changes.md#short-versions-to-see-basic-flow-and-considerations

**Review by dsasser — APPROVED** [review]
Current changes look good, though to @swirtSJW's point, we might want to hide new fields so editors are not confused.

**Comment by chri5tia** [comment]
If there are any editor docs that need to be updated as a result of this work, the workflow for this in relation to the PRs are as follows:

https://github.com/department-of-veterans-affairs/va.gov-cms/blob/b9826842a90e7cf0f60101be8746e0c422bd85b9/docroot/modules/custom/va_gov_db/va_gov_db.install#L239

**Comment by chri5tia** [comment]
After talking with Steve, I decided it would be good for myself and the team to walk through the best practice outlined earlier by him.

I am modifying the code in this PR to hide the description field so that it is in the schema but not usable yet by editors. Then the PR will be ready for review.

The next step would be for the FE PR to sync to production and then I would create a new PR with the code changes to make the description field available.

**Comment by chri5tia** [comment]
@swirtSJW I believe the associated PR from content build is dependent on this going to testing and staging in order for further testing by @jtmst. Is there a different content build piece you are referring to?

**Comment by swirtSJW** [comment]
THis all seems to work fine but seems clunky in the supportive elements:
1) The supportive text speaks an old truth but a new lie.
image

2. Odd grouping under other and nothing gives the impression that this is only used to collect existing content.
image

3. Another screen with no indication that I can only choose from existing QAs
image

**Comment by swirtSJW** [comment]
> @swirtSJW I believe the associated PR from content build is dependent on this going to testing and staging in order for further testing by @jtmst. Is there a different content build piece you are referring to?

Sorry, no that is not the desired workflow.  We went  out of our way to make it so that tugboat environments can build a copy of the FE using a PR from content-build, so all this can be tested before either PR hits prod.

**Comment by chri5tia** [comment]
Regarding the help text comments, I made a rough draft of fixes for UX review, if that's helpful:

```
diff --git a/config/sync/field.field.paragraph.q_a_group.field_q_as.yml b/config/sync/field.field.paragraph.q_a_group.field_q_as.yml
index ecfb95e9..2959c32b 100644
--- a/config/sync/field.field.paragraph.q_a_group.field_q_as.yml
+++ b/config/sync/field.field.paragraph.q_a_group.field_q_as.yml
@@ -8,21 +8,16 @@ dependencies:
     - paragraphs.paragraphs_type.q_a_group
   module:
     - entity_reference_validators
-    - epp
 third_party_settings:
   entity_reference_validators:
     circular_reference: false
-    circular_reference_deep: false
     duplicate_reference: false
-  epp:
-    value: ''
-    on_update: 0
 id: paragraph.q_a_group.field_q_as
 field_name: field_q_as
 entity_type: paragraph
 bundle: q_a_group
 label: 'Q&As'
-description: 'Add single or multiple Q&As. Reordering of Q&A is available via drag and drop. Q&As made available here are created using the Q&A - single content type.'
+description: 'Add single or multiple Q&As. Reordering of Q&A is available via drag and drop.'
 required: true
 translatable: false
 default_value: {  }
diff --git a/config/sync/paragraphs.paragraphs_type.q_a_group.yml b/config/sync/paragraphs.paragraphs_type.q_a_group.yml
index 6e19eb81..792cc74e 100644
--- a/config/sync/paragraphs.paragraphs_type.q_a_group.yml
+++ b/config/sync/paragraphs.paragraphs_type.q_a_group.yml
@@ -11,5 +11,5 @@ id: q_a_group
 label: 'Q&A group'
 icon_uuid: null
 icon_default: null
-description: 'For content formatted as a series of reusable questions and answers from existing Q&As using the Q&A - Single content type. Use this (instead of Rich text) for better accessibility and easy rearranging. Currently available on FAQ Page content type and Resources and Support content types.'
+description: 'For content formatted as a series of questions and answers in "FAQ - multiple Q&A" content type. Use this (instead of Rich text) for better accessibility and easy rearranging.'
```

**Comment by jilladams** [comment]
UI notes added to https://github.com/department-of-veterans-affairs/va.gov-cms/issues/10147 for follow up

**Comment by chri5tia** [comment]
We are going to make the UX help text copy changes in #10147 in a separate body of work once the copy is signed off on. Timeline to get this out outweighs UX considerations for future users. But since #10147 is a small lift I should be able to fit it into the next sprint.

This PR is holding up the templating work that Josh needs to do for the front end piece. It is ready to review and merge.

**Comment by swirtSJW** [comment]
Adding links to slack discussions here
- [design related concerns](https://dsva.slack.com/archives/C52CL1PKQ/p1660231284396329) (tabled for new ticket)
- [End to end testing with FE PR in tugboat](https://dsva.slack.com/archives/C52CL1PKQ/p1660226876441269)

The code  looks good can be merged once tugboat E2E is performed.  Not sure if that is done as I don't see reference to it here

**Review by swirtSJW — APPROVED** [review]

**Comment by chri5tia** [comment]
The work here is done and approved. What we need to do is test end to end in tug boat. In order to do that, [the front end PR](https://github.com/department-of-veterans-affairs/content-build/pull/1253) needs to go out. He is running up against tugboat build issues that need further troubleshooting.

https://cms-zqis1amj7i3jmmqvydvdxfmptbsn7zzf.ci.cms.va.gov/sites/default/files/build.txt

**Comment by chri5tia** [comment]
This branch should wait for the FE component before going out to production.

**Comment by dsasser** [comment]
@chri5tia Jut a point of clarity: we are waiting for the FE PR to be in a testable state (not released, however), thereby allowing us to do the E2E testing you mentioned, which is the final blocker for both this and the FE PR.

**Review by dsasser — APPROVED** [review]

**Review by jtmst — APPROVED** [review]
